### PR TITLE
Issue 12421 - Allow simpler syntax for lambda template declarations

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=12421

### CODE REVIEW

**Comment by 9rnsr** [comment]
Rebased.

**Comment by MartinNowak** [comment]
I'm still in favor of this change as named variadic lamdas are useful, e.g. to factor out the predicate of an algorithm.
```d
alias header = v => v.startsWith('HTTP');
environment.filter!header.each!writeln;
```

**Comment by Hackerpilot** [comment]
I can see that this is a useful improvement when allowed as an "auto" declaration, but I'm not convinced that we should be allowing "alias" here. The grammar for "alias" says that alias is something that should be used for types: http://dlang.org/grammar.html#AliasDeclaration

I know that you can also alias symbols (if they can also parse as types), but I see that as a problem with the specification, (because it should allow symbols), and the parser (because the parser requires that things being aliased parse as types).

If we want `alias` to work for expressions, that should be a DIP.

**Comment by MartinNowak** [comment]
Also see https://issues.dlang.org/show_bug.cgi?id=9426#c3.

**Comment by yebblies** [comment]
Could you split out the refactorings into their own pull requests?

**Comment by 9rnsr** [comment]
@yebblies Done: #4663.

**Comment by 9rnsr** [comment]
Rebase onto the latest master. Now this PR contains only one commit!

**Comment by MartinNowak** [comment]
> I can see that this is a useful improvement when allowed as an "auto" declaration, but I'm not convinced that we should be allowing "alias" here.

It's still a template. You can alias a template but you cannot assign it to a variable, so alias makes sense auto doesn't.

**Comment by 9rnsr** [comment]
Rebased on master, and ready for ddmd.

**Comment by ZombineDev** [comment]
LGTM

**Comment by yebblies** [comment]
This looks good, and I plan to merge it soon unless there are objections.  I think this is a straightforward addition of missing cases to an existing feature.

File: test/runnable/funclit.d [hunk 1]
```diff
@@ -1069,6 +1069,60 @@ void test11774()
...
+
+    test!(q{ delegate     (    x){ return x; } });
+    test!(q{ delegate     (int x){ return x; } }, false);
+    test!(q{ delegate int (    x){ return x; } }, false);
+    test!(q{ delegate int (int x){ return x; } }, false);
+
+    // This is problematic case, and should be disallowed in the future.
+    alias f = x => y;
+    int y = 10;
+    assert(f(1) == 10);;
```
Comment (yebblies) [inline test/runnable/funclit.d]: We do already maintain a linked list of variables to allow goto-skips-init errors, something similar could be used to fix this.

**Comment by MetaLang** [comment]
+10

**Comment by yebblies** [comment]
Auto-merge toggled on

**Comment by 9rnsr** [comment]
Thanks.

File: test/runnable/funclit.d (hunk 1 above)
Comment (denis-sh) [inline test/runnable/funclit.d]: Looks like `;;` isn't intentional here.

**Comment by MartinNowak** [comment]
A changelog notice with a small example would be preferred.

**Comment by ZombineDev** [comment]
@9rnsr @MartinNowak @yebblies @Hackerpilot this was good enhancement, however it is much less useful compared to the lamda templates supported by C++14. Can you please join me in discussing the pros & cons of the design on the forum: http://forum.dlang.org/post/xfjoktyjtoojhszrlugi@forum.dlang.org?

**Pull request merged**